### PR TITLE
Improve async rendering section of chapter 12

### PR DESCRIPTION
**Pull request by chrishtr**
This PR does the following:

* Improves description of the what and why of the code
* Breaks up the code more with explanatory text
* Fixes a bug where `run_rendering_pipeline` was called on the timer thread, and replaces it with a call on the event loop.

### CODE REVIEW

File: book/scheduling-and-threading.md [hunk 1]
```diff
@@ -174,18 +174,29 @@ Asynchronous rendering
...
+class, which takes two parameters: a time delta from now, and a function to
+call when that time expires. It will start a new thread and call that
+function *on that thread*[^thread-timer] at the desired time. When the
+function completes, the timer thread is automatically ended.
+
+[^thread-timer]: This is convenient because it enables the timer to
+execute at about the time desired---even if the current thread is busy at that
+time---but, as we'll see, leads to some complications if you really want the
+function to be called on the same thread as the one calling `set_timeout`.
+But in any case, it wouldn't make any sense to run the task on the "current"
```
Comment (pavpanchekha) [inline book/scheduling-and-threading.md:196]: I found this footnote really confusing. The actual point it's making is pretty clear, so we should frame it positively (here's what's true) instead of negatively (here's why that other belief is false).

File: book/scheduling-and-threading.md [hunk 2]
```diff
@@ -195,34 +206,53 @@ def set_timeout(func, sec):
...
-Also, rename `render` to `run_rendering_pipeline`, and add a new
-`run_animation_frame` method that runs the pipeline and calls the other
-rendering pipeline stages on `Browser` via a new `raster_and_draw`
-method (which we'll implement shortly).
+Next, add three dirty bits to `Tab`:
+
+* `needs_pipeline_update`, indicating that
+the pipeline needs to be re-run.
+* `display_scheduled`, indicating that `set_timeout` was already called (this
+avoids double-running `set_timeout` unnecessarily).
```
Comment (pavpanchekha) [inline book/scheduling-and-threading.md:214]: I'm a little confused:

1) Why it's called `display_scheduled`, if it's about `set_timeout`
2) What does this mean by double-running `set_timeout`?
3) How these flags relate to each other.

Comment (pavpanchekha) [reply book/scheduling-and-threading.md:214]: More precisely, I think this goes back to my concern that there are a lot of flags and we haven't given the reader any way of thinking about them. Why so many flags? How do you remember how they relate? How do you choose to have more or fewer? I'm not actually sure if we want these dirty flags to be a major player in this chapter, but if we do we should talk more about managing them.

File: book/scheduling-and-threading.md [hunk 3]
```diff
@@ -195,34 +206,53 @@ def set_timeout(func, sec):
...
+* `run_pipeline_now`, indicating that the event loop should
+run the pipeline.
+
+Add methods to set the dirty bits and call `set_timeout` as
+needed.[^race-condition]
+
+[^race-condition]: This code has a race condition because it's setting variables
+on one thread from another without a thread lock. Let's ignore that for now and
+fix it later in the chapter.
```
Comment (pavpanchekha) [inline book/scheduling-and-threading.md:223]: Potentially we should move discussion of locks way up to where we first introduce threads (setTimeout) so that we don't have to continuously apologize for it throughout the book.

File: book/scheduling-and-threading.md [hunk 4]
```diff
@@ -195,34 +206,53 @@ def set_timeout(func, sec):
...
 
         if not self.display_scheduled:
             set_timeout(callback, REFRESH_RATE_SEC)
             self.display_scheduled = True
+```
+ 
+Also, rename `render` to `run_rendering_pipeline`, and add a new
+`run_animation_frame` method that runs the pipeline and calls the other
+rendering pipeline stages on `Browser` via a new `raster_and_draw`
+method.
```
Comment (pavpanchekha) [inline book/scheduling-and-threading.md:250]: This is much better.

File: book/scheduling-and-threading.md [hunk 5]
```diff
@@ -237,10 +267,33 @@ class Tab:
...
+
+``` {.python expected=False}
+    while True:
+        if sdl2.SDL_PollEvent(ctypes.byref(event)) != 0:
+            # ...
+        active_tab = browser.tabs[browser.active_tab]
+        if active_tab.run_pipeline_now:
+            active_tab.run_pipeline_now = False
+            active_tab.run_animation_frame()
+```
```
Comment (pavpanchekha) [inline book/scheduling-and-threading.md:291]: Is here the right place to explain why we must set `run_pipeline_now` before running the animation frame? It is related somehow to the idea that you need to be able to schedule a rAF inside a rAF callback maybe...

**Review by pavpanchekha — APPROVED** [review]

**Comment by pavpanchekha** [comment]
I think my main concern with this PR and also the previous one on setTimeout is that we now have a lot of flags. We should simplify, organize, or explain. I don't think that should block merging, however.

File: book/scheduling-and-threading.md [hunk 6]
```diff
@@ -174,18 +174,29 @@ Asynchronous rendering
...
-`threading.Timer` class, which takes two parameters: a time delta from now, and
-a function to call when that time expires.
+This function will run a callback at a specified time in the future. You can do
+that by starting a new [Python thread][python-thread] via the `threading.Timer`
+class, which takes two parameters: a time delta from now, and a function to
+call when that time expires. It will start a new thread and call that
+function *on that thread*[^thread-timer] at the desired time. When the
+function completes, the timer thread is automatically ended.
+
+[^thread-timer]: This is convenient because it enables the timer to
```
[the remark this answers is not in the document]
Comment (chrishtr) [reply book/scheduling-and-threading.md:192]: Yes-- see line 189

File: book/scheduling-and-threading.md (hunk 1 above)
Comment (chrishtr) [reply book/scheduling-and-threading.md:196]: Fixed in another PR.

File: book/scheduling-and-threading.md (hunk 2 above)
Comment (chrishtr) [reply book/scheduling-and-threading.md:214]: display_scheduled avoids running stuff twice, but doesn't mean the event loop should run just now. needs_animation_frame means "please run it at the next opportunity"

display_scheduled avoids double running because if you call set_timeout twice, before the callback has run, the second-time will be a no-op.

Unfortunately I do think we need at least some of these dirty flags or the browser will not function correctly (e.g. the ones above). Other flags are optional and for performance reasons. We could push them to the end of the chapter.

File: book/scheduling-and-threading.md (hunk 3 above)
Comment (chrishtr) [reply book/scheduling-and-threading.md:223]: ok, good point

**Comment by chrishtr** [comment]
TL;DR I agree that we should attempt to simplify and keep improving. Will keep at it.

File: book/scheduling-and-threading.md (hunk 2 above)
Comment (pavpanchekha) [reply book/scheduling-and-threading.md:396]: Agreed, sounds good.
